src/logging.rs

```diff
@@ -82,16 +83,29 @@ fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
         data.push_str(&field_payload);
     }
 }
-
-/// Print a message to the journal with the given priority.
+/// Send a message with structured properties to the journal.
 ///
-/// This uses the default systemd-journald socket.
-pub fn journal_print(priority: Priority, msg: &str) -> Result<()> {
+/// vars is a map of journald fields to values.This function uses the default systemd-journald socket.
+/// Iterator may contains `PRIORITY` or `MESSAGE` fields. Instead of duplicating those fields,we'll use the ones which are already injected through journal.send(). So there's a condition in the iterator to skip those two values if present in the map.
+pub fn journal_send<K, V>(
+    priority: Priority,
+    msg: &str,
+    vars: impl Iterator<Item = (K, V)>,
+) -> Result<()>
+where
+    K: AsRef<str>,
+    V: AsRef<str>,
```

Comment:
This is OK, though... if we want to truly match the API semantics, the value could technically just be bytes. UTF-8 is recommended, but binary is fully supported. So this could even be `AsRef<[u8]>`.

Reply:
It looks like both the C and the Go API handle strings, and even binary stuff like `MESSAGE_ID` is hex-encoded. I think that handling strings here is ok.

Reply:
The protocol itself allows for any arbitrary binary data. `systemd-coredump` for example can be configured to store full coredump in the journal (which is why journal messages have a crazy limit like 768M). And in rpm-ostree, I almost made use of binary values as well in the journal when implementing https://github.com/projectatomic/rpm-ostree/pull/1813.

Reply:
(To be clear, this WFM as is too, just wanted to flag we weren't exposing the full capabilities of the API. We can always add a separate API later on if the need arises.)

Reply:
Yeah, it is starting to get tricky. The C API has a further split between `_send` and `_sendv` for the string-vs-bytes topic.
I'd say we keep this signature for the moment and then iterate on a lower level helper on the side. I think in that case having split priority/message does not make sense either.